radareorg/cutter
# Show in action and refactoring

**Your checklist for this pull request**


**Detailed description**
This PR implements the following
1. Refactor `actionRenameThingHere` for using the refactored `RCodeAnnotation`. (It now contains one struct inside union named `reference` for annotations of type constant variable, global variable, and function name
2. Implement the basic show in action that shows the decompiled line in other widgets(disassembly, graph, ..)
3. Implement targeted show in menu for functions, global variables, and constant variables
4. Implement add name, rename name, remove name actions for global and constant variables

Actions for global and constant variables 
![renameaddremovename](https://user-images.githubusercontent.com/18501167/87765466-792c7500-c835-11ea-84f1-4271bb493ebb.gif)
Show in action
![showinaction](https://user-images.githubusercontent.com/18501167/87765487-7f225600-c835-11ea-9086-6e7d8dbcfced.gif)



**Test plan (required)**
**Note**: CI will fail because the corresponding PRs in r2 and r2ghidra-dec haven't been merged yet.


1. Fetch PR [#124](https://github.com/radareorg/r2ghidra-dec/pull/124) from r2ghidra-dec before compiling this PR.
2. Check if the rename function action is working as expected after refactoring.
3. Check if the show in action and targeted show in action are working as expected. Suggest a _better text_ for the `targeted show in` action that's shown.
4. Check if `add name,` `remove name`, and `rename name` actions are working perfectly.
5. Check code.

**Detailed Test Plan For actions with annotations for variables**
Download this binary: 
[dectest32.zip](https://github.com/radareorg/cutter/files/4960680/dectest32.zip)
Load in Cutter
Go to function `sym.get_global_array_entry`
You shall see an unnamed global variable that's used there.

_Test for add name/rename/remove name actions for global variables_

- [ ] Right-click on the address(that refer to the global variable) -> Add name -> Give a name. It should now show the name that you gave instead of `*(undefined4 *)0x804c034`
- [ ] Click on that variable -> Shift+N (shortcut) -> It should now show the rename dialog, give a new name and see if it's changing as it should.
- [ ] Right-click on the variable -> Click on `Remove name` -> The name should get removed and you should be seeing `*(undefined4 *)0x804c034`.

_Test for show-in action for global variables_

- [ ] Right-click on the variable -> You should be seeing a menu with text: `Show 0x804c034(addr here)`.
- [ ] Go to that menu and click various options and make sure they are opening new widgets for the offset `0x804c034`.
- [ ] Go back to the decompiler and add a name to the variable.
- [ ] Right-click on the variable -> You should be seeing a menu with text `Show <name of variable> in` instead of `Show 0x804c034(addr here)`.
- [ ] Go to the menu and click on the different actions and make sure all of them are working like how the corresponding `show at address` in the disassembly works.

_Test for show-in action for constant variables that are annotated_

- [ ] Go to function `sym.PrintAmbassador`
- [ ] Right-click `pure` in the line `sym.imp.printf("pure");`.
- [ ] You should not be seeing any rename/add/remove name actions.
- [ ] You should be seeing a menu named `Show str.pure in`
- [ ] Go into the menu and test all the actions and make sure they are working like how you tested show-in menu actions for the global variables.



**Closing issues**
